### PR TITLE
(Feat) Beat - Breatout host function: create and cleanSessions - SPARK-400358

### PR DESCRIPTION
**Pull request by JudyZhuHz**
# COMPLETES #<  Breatout host function: create and cleanSessions>
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-400358?filter=-1

## This pull request addresses

// Deletes a breakout session
breakout.delete()

// Create a breakout session
breakouts.create(sessions)

## by making the following changes

Add two breakout function for host

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >
Add Test case
Run the demo page, see the native behavior
Click "creatonesession" button, open the native breakout button, display one breakout session item

### I certified that

- [x ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ x] I discussed changes with code owners prior to submitting this pull request

- [ x] I have not skipped any automated checks
- [ x] All existing and new tests passed
- [ x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/breakouts/README.md [hunk 1]
```diff
@@ -119,8 +119,11 @@ breakout.remove(participant)
 // This can only be done when breakouts are inactive
 breakout.delete()
 
-// Create a breakout session
-breakouts.create(sessionName)
+// Create breakout sessions
+breakouts.create(sessions)
+
+// Delete all breakout sessions
+breakouts.clearSessions()
```
Comment (arun3528) [inline packages/@webex/plugin-meetings/src/breakouts/README.md:126]: arent we calling them as rooms / sessions / meetings 

sessions looks little odd here

**Review by arun3528 — COMMENTED** [review]
i think we should not use the word sessions here any thoughts ?

at least the developer facing stuff

File: packages/@webex/plugin-meetings/src/breakouts/README.md (hunk 1 above)
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/src/breakouts/README.md:126]: https://sqbu-github.cisco.com/pages/WebExSquared/breakout/api/resource/createBreakoutSessions.html

I checked the design doc,  call the breakout as sessions.

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 2]
```diff
@@ -460,5 +460,63 @@ describe('plugin-meetings', () => {
...
+              },
+            ],
+          }
+        });
+
+        assert.equal(breakouts.groups[0].status, "CLOSE")
+      });
+    });
+
+    describe('creat', () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:493]: typo on create

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 3]
```diff
@@ -460,5 +460,63 @@ describe('plugin-meetings', () => {
...
+      it('repose not include groups info', async () => {
+        const sessions = [{'name':'session1', "anyoneCanJoin" : true}];
+        const result = await breakouts.create(sessions);
+
+        assert.equal(result, 'REQUEST_RETURN_VALUE');
+
+      });
+
+      it('repose include groups info', async () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:502]: type repose -> response?

File: packages/@webex/plugin-meetings/src/breakouts/README.md [hunk 4]
```diff
@@ -119,8 +119,11 @@ breakout.remove(participant)
 // This can only be done when breakouts are inactive
 breakout.delete()
 
-// Create a breakout session
-breakouts.create(sessionName)
+// Create breakout sessions
+breakouts.create(sessions)
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/breakouts/README.md:123]: we need a better description here on what a session is i.e. the argument type of sessions

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 5]
```diff
@@ -348,6 +349,57 @@ const Breakouts = WebexPlugin.extend({
...
   },
+
+  /**
+   * Create new breakout sessions
+   * @param {object} sessions -- breakout session group
+   * @returns {Promise}
+   */
+  async create(sessions) {
+    // @ts-ignore
+    const breakInfo = await this.webex.request({
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:360]: Both the create and the delete can fail if someone else is editing at the same time. The server response with 
```
{"errorCode":201409024,"message":"Edit lock token mismatch","errors":[{"errorCode":201409024,"description":"Edit lock token mismatch"}],"trackingId":"webex-js-sdk_7b6a0918-5eaf-493e-a989-40073c2b38ea_56"}
```

we should handle this case and return an error

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts (hunk 3 above)
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts]: Fixed.

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts (hunk 2 above)
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts]: Fixed

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 5 above)
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts]: add catch and throw error

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 6]
```diff
@@ -460,5 +460,93 @@ describe('plugin-meetings', () => {
...
+                action: BREAKOUTS.ACTION_TYPES.DELETE,
+              },
+            ],
+          }
+        });
+
+        assert.equal(breakouts.groups[0].status, "CLOSE")
+      });
+
+      it('response include erorr info', async () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:492]: typo erorr -> error

File: packages/@webex/plugin-meetings/src/breakouts/README.md (hunk 4 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/README.md:124]: Do we know what options are available here?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 7]
```diff
@@ -348,6 +350,77 @@ const Breakouts = WebexPlugin.extend({
...
+              sessions,
+            },
+          ],
+        },
+      })
+      .catch((error) => {
+        if (error.body && error.body.errorCode === BREAKOUTS.ERROR_CODE.EDIT_LOCK_TOKEN_MISMATCH) {
+          LoggerProxy.logger.info(`Breakouts#create --> Edit lock token mismatch`);
+
+          return Promise.reject(new BreakoutEditLockedError('Edit lock token mismatc', error));
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:377]: typo mismatc-> mismatch

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 8]
```diff
@@ -460,5 +460,93 @@ describe('plugin-meetings', () => {
...
+            ],
+          }
+        });
+
+        assert.equal(breakouts.groups[0].status, "CLOSE")
+      });
+
+      it('response include error info', async () => {
+        webex.request.returns(Promise.resolve({
+          body: {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:494]: see comment on the create test about rejecting this promise

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 9]
```diff
@@ -460,5 +460,93 @@ describe('plugin-meetings', () => {
...
+        const sessions = [{'name':'session1', "anyoneCanJoin" : true}];
+
+        webex.request.returns(Promise.resolve({
+          body: {
+            "errorCode":BREAKOUTS.ERROR_CODE.EDIT_LOCK_TOKEN_MISMATCH,
+            "message":"Edit lock token mismatch"
+          }
+        }));
+
+        const error = await breakouts.create(sessions);
```
[the remark this answers is not in the document]
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:545]: Yes, i add the "rejects when edit lock token mismatch" case below

File: packages/@webex/plugin-meetings/src/breakouts/README.md (hunk 4 above)
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/src/breakouts/README.md:124]: As locus design:
"sessions" : [ {
      "name" : "Breakout Session 1",
      "subConfId" : 1,
      "anyoneCanJoin" : false,
      "id" : "ea2db6e1-2da9-4f72-9d62-f099e5e3c7b4"
    }, {
      "name" : "Breakout Session 2",
      "anyoneCanJoin" : true
    } ]

**Review by Coread — APPROVED** [review]

**Review by arun3528 — APPROVED** [review]

**Pull request merged**